iree/compiler/Dialect/Flow/Transforms/Passes.cpp

```diff
@@ -116,59 +116,63 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
                                     const TransformOptions &transformOptions) {
   // Special case peephole optimizations.
   FunctionLikeNest(passManager)
-      .addPass(createConvertConv2D1x1ToMatmulPass)
+      .addPass(IREE::Flow::createConvertConv2D1x1ToMatmulPass)
       .addPredicatedPass(clEnableConvToImg2Col,
-                         createConvertConv2DToImg2ColPass)
+                         IREE::Flow::createConvertConv2DToImg2ColPass)
       // Pad linalg op
-      .addPredicatedPass(clEnablePaddingLinalgOps,
-                         []() {
-                           return createPadLinalgOpsToIntegerMultiplePass(
-                               clLinalgOpsPaddingSize);
-                         })
+      .addPredicatedPass(
+          clEnablePaddingLinalgOps,
+          []() {
+            return IREE::Flow::createPadLinalgOpsToIntegerMultiplePass(
+                clLinalgOpsPaddingSize);
+          })
 
       // Input should now be legal.
-      .addPass(createVerifyInputLegalityPass);
+      .addPass(IREE::Flow::createVerifyInputLegalityPass);
 
   passManager.addPass(mlir::createLinalgNamedOpConversionPass());
+
+  // Expand tensor shapes into SSA values and optimize the whole program.
+  // The more we are able to equate shape dimensions at this level the better
+  // our fusions will be.
+  passManager.addPass(IREE::Flow::createExpandTensorShapesPass());
```

Comment:
Is there any integration / e2e test that this enable support for, or it this just an optimization? Should we have [more] programs in our benchmark suite that include dynamic shapes?

Reply:
Yes, this enables me to write e2e tests that I can't write so far, specifically about any e2e test involving functions taking a dynamic-shapes tensor as argument and performing a `tensor.dim` on it.

Reply:
Here's a e2e test that could be added to `iree/test/e2e/regression` (failing without this PR, succeeding with this PR):

```
func private @foo(%lhs : tensor<?x?xi8>) -> tensor<?xi32> {
  %c_0_index = arith.constant 0 : index
  %c_1_index = arith.constant 1 : index
  %m_size = tensor.dim %lhs, %c_0_index : tensor<?x?xi8>
  %c_0 = arith.constant 0 : i32

  // compute the sums along rows of %lhs.
  %lhs_i32 = arith.extsi %lhs : tensor<?x?xi8> to tensor<?x?xi32>
  %init_lhs_sums_uninitialized = linalg.init_tensor [%m_size] : tensor<?xi32>
  %zero_lhs_sums = linalg.fill(%c_0, %init_lhs_sums_uninitialized) : i32, tensor<?xi32> -> tensor<?xi32>
  return %zero_lhs_sums : tensor<?xi32>
}

func @bar(%lhs : tensor<?x?xi8>) {
    %a = call @foo(%lhs) : (tensor<?x?xi8>) -> tensor<?xi32>
    check.expect_eq(%a, %a) : tensor<?xi32>
    return
}
```

Reply:
I tried hard to provide a minimized testcase, but the best I came up with triggers a null deref (with this PR applied):

```
func private @reduce(%matrix : tensor<?x?xi8>) -> tensor<?xi32> {
  %c_0_index = arith.constant 0 : index
  %c_1_index = arith.constant 1 : index
  %m_size = tensor.dim %matrix, %c_0_index : tensor<?x?xi8>
  %c_0 = arith.constant 0 : i32

  %matrix_i32 = arith.extsi %matrix : tensor<?x?xi8> to tensor<?x?xi32>
  %init_matrix_uninitialized = linalg.init_tensor [%m_size] : tensor<?xi32>
  %zero_matrix = linalg.fill(%c_0, %init_matrix_uninitialized) : i32, tensor<?xi32> -> tensor<?xi32>
  return %zero_matrix : tensor<?xi32>
}

func private @testcase(%matrix : tensor<?x?xi8>) {
  %zeros = util.unfoldable_constant dense<[0, 0, 0]> : tensor<3xi32>
  %zeros_dyn = tensor.cast %zeros : tensor<3xi32> to tensor<?xi32>
  %a = call @reduce(%matrix) : (tensor<?x?xi8>) -> tensor<?xi32>
  check.expect_eq(%a, %zeros_dyn) : tensor<?xi32>
  return
}

func @test() {
  %matrix = util.unfoldable_constant dense<[
      [1, 2, 3, 4],
      [5, 6, 7, 8],
      [9, 10, 11, 12]]> : tensor<3x4xi8>
  %matrix_dyn = tensor.cast %matrix : tensor<3x4xi8> to tensor<?x?xi8>
  call @testcase(%matrix_dyn) : (tensor<?x?xi8>) -> ()
  return
}
```

```
(gdb) bt
#0  0x000000000c886e17 in mlir::detail::IROperandBase::insertInto<mlir::IRObjectWithUseList<mlir::OpOperand> >
    (this=0x6130002d3210, useList=0x0)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:82
#1  0x000000000c886d49 in mlir::IROperand<mlir::OpOperand, mlir::Value>::insertIntoCurrent (
    this=0x6130002d3210)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:159
#2  0x000000002120b1ad in mlir::IROperand<mlir::OpOperand, mlir::Value>::IROperand (this=0x6130002d3210, 
    owner=0x6130002d3180, value=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:115
#3  0x00000000212077fb in mlir::OpOperand::IROperand (this=0x6130002d3210)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/Value.h:263
#4  0x00000000211f5d87 in mlir::detail::OperandStorage::OperandStorage (this=0x6130002d31c0, 
    owner=0x6130002d3180, trailingOperands=0x6130002d31d0, values=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/OperationSupport.cpp:227
#5  0x00000000211be072 in mlir::Operation::create (location=..., name=..., resultTypes=..., operands=..., 
    attributes=..., successors=..., numRegions=0)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:113
#6  0x00000000211bef0f in mlir::Operation::create (location=..., name=..., resultTypes=..., operands=..., 
    attributes=..., successors=..., regions=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:49
#7  0x00000000211be767 in mlir::Operation::create (state=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:38
#8  0x0000000020e00aa1 in mlir::OpBuilder::createOperation (this=0x7fffffff7780, state=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Builders.cpp:381
#9  0x0000000014f6b1d8 in mlir::OpBuilder::create<mlir::CallOp, llvm::StringRef, llvm::SmallVector<mlir::Type, 6u>&, llvm::SmallVector<mlir::Value, 6u>&> (this=0x7fffffff7780, location=..., args=..., args=..., args=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/llvm/../mlir/include/mlir/IR/Builders.h:434
#10 0x000000001680aa83 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandCallOp (op=..., 
    indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:395
#11 0x0000000016802503 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandSubviews (
    op=0x6110000e49c0, globalMap=..., indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:490
#12 0x0000000016811b1a in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandRegion (region=..., 
    globalMap=..., indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:250
#13 0x000000001680a045 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandFuncOp (op=..., 
    globalMap=..., indexSet=..., subviewMap=...)
```

Reply:
thx - fixed